### PR TITLE
[logging] Install Fluentd on Windows instance

### PR DESCRIPTION
**Pull request by suhanime**
This commit sets up an Ansible Playbook to install Ruby Development Kit
and Fluentd on the Windows instance to enable logging.

### CODE REVIEW

File: ansible/install_td_agent.yaml [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+  tasks:
+    - name: Create temporary target directory
+      win_tempfile:
+        state: directory
+      register: tempdir_win
+    - name: Copy td-agent msi
+      win_copy:
+        src: "{{ hostvars['localhost']['tempdir']['path'] }}/td-agent.msi"
+        dest: "{{ tempdir_win.path }}\\"
+    - name: Run td-agent msi
```
Comment (aravindhp) [inline ansible/install_td_agent.yaml]: I see the `when` clause being used in #21. Shouldn't that be used here too? I would like to have consistency regarding such things across our playbooks.

File: ansible/tasks/logging/README.md [hunk 2]
```diff
@@ -0,0 +1,34 @@
+# Playbooks to install FluentD and FluentD plugins
```
Comment (aravindhp) [inline ansible/tasks/logging/README.md]: There is a lot of overlap with this README and the one introduced in #21. Please work together and ensure that information is not repeated in multiple places. Ideally there should be a main README with common info and links to the individual READMEs for different playbooks.

File: ansible/tasks/logging/install_fluentd.yaml [hunk 3]
```diff
@@ -0,0 +1,24 @@
+---
+- name: Create temporary directory on windows host
+  win_tempfile:
+    state: directory
+  register: tempdir_win
+- name: Fetch ruby devkit exe
+  win_get_url:
+    url: "{{ ruby_fetch_url }}"
+    dest: "{{ tempdir_win.path }}\\ruby_devkit.exe"
+- name: Run ruby devkit exe
```
Comment (aravindhp) [inline ansible/tasks/logging/install_fluentd.yaml]: I see the `when` clause being used in #21. Shouldn't that be used here too? I would like to have consistency regarding such things across our playbooks.

File: ansible/tasks/logging/main.yml [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+- hosts: windows
+  vars:
+    ruby_version: "RubyInstaller-2.4.7-1/rubyinstaller-devkit-2.4.7-1-x86.exe"
+    ruby_installer_path: "https://github.com/oneclick/rubyinstaller2/releases/download"
+    ruby_fetch_url: "{{ ruby_installer_path }}/{{ ruby_version }}"
+  tasks:
+   - name: Logging | Install Fluentd
+     include_tasks: install_fluentd.yaml
+   - name: Logging | Install Fluentd plugins
+     include_tasks: install_plugins.yml
```
Comment (aravindhp) [inline ansible/tasks/logging/main.yml]: I don't see this file. I assume this is being added in #39. Won't it cause an error if we try running it?

File: ansible/tasks/logging/install_fluentd.yaml [hunk 5]
```diff
@@ -0,0 +1,24 @@
+---
+- name: Create temporary directory on windows host
+  win_tempfile:
+    state: directory
+  register: tempdir_win
+- name: Fetch ruby devkit exe
+  win_get_url:
+    url: "{{ ruby_fetch_url }}"
+    dest: "{{ tempdir_win.path }}\\ruby_devkit.exe"
```
Comment (aravindhp) [inline ansible/tasks/logging/install_fluentd.yaml]: What is the reason to rename the exe? Can't the variable from `main.yml` just be reused here. It will help if we have to log into nodes to debug to have the version in the filename.

File: ansible/tasks/logging/install_fluentd.yaml [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+- name: Run ruby devkit exe
+  win_package:
+    path: "{{ tempdir_win.path }}\\ruby_devkit.exe"
+    creates_path: "C:\\Ruby26-x64\\bin\\ruby.exe"
```
Comment (aravindhp) [inline ansible/tasks/logging/install_fluentd.yaml]: `rubyinstaller-devkit-2.4.7-1-x86.exe` implies it is a 32-bit binary. But the path you are creating implies 64-bit. Why?

Comment (suhanime) [reply ansible/tasks/logging/install_fluentd.yaml]: I missed the change here, I will take care of it. However, this path would be version dependent, should we want to change the version of devkit in `main.yml`, so I will make the path here a variable.
I am not creating a path specifically - this is a requirement for win_package module so it won't re-install if already installed. More information found [here](https://docs.ansible.com/ansible/latest/modules/win_package_module.html#parameters)

File: ansible/tasks/logging/README.md [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+```
+$ sudo dnf install libselinux-python
+$ pip install ansible pywinrm
+```
+- A Windows Server 2019 running in the same VPC as the OpenShift cluster with winrm enabled.
+You can use `wni` tool to setup the Windows node following this [link](https://github.com/openshift/windows-machine-config-operator/tree/master/tools/windows-node-installer).
+The winrm module is enabled automatically if you use `wni` tool to create the server node.
+- Login to your cluster using `oc` command.
+
+Modify the `hosts` file with the instance information.
```
Comment (ravisantoshgudimetla) [inline ansible/tasks/logging/README.md]: Can you overwrite the variables that you're having in the hosts file from command line?

File: ansible/tasks/logging/README.md [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+- Set up an OpenShift cluster
+- Ensure Ansible environment is set up to run the playbook. For a Fedora user, sample steps are:
+```
+$ sudo dnf install libselinux-python
+$ pip install ansible pywinrm
+```
+- A Windows Server 2019 running in the same VPC as the OpenShift cluster with winrm enabled.
+You can use `wni` tool to setup the Windows node following this [link](https://github.com/openshift/windows-machine-config-operator/tree/master/tools/windows-node-installer).
+The winrm module is enabled automatically if you use `wni` tool to create the server node.
+- Login to your cluster using `oc` command.
```
Comment (ravisantoshgudimetla) [inline ansible/tasks/logging/README.md]: Why should we login? Are we using any subsequent `oc` commands in the ansible playbook?

File: ansible/tasks/logging/install_fluentd.yaml (hunk 6 above)
Comment (ravisantoshgudimetla) [reply ansible/tasks/logging/install_fluentd.yaml]: What is the default behaviour here? If we don't use `creates_path`, would the installation happen at `C:\<product_name>`? If yes, I am in favour of that

File: ansible/tasks/logging/install_fluentd.yaml [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+    state: present
+    arguments:
+    - /VERYSILENT
+    - /SUPPRESSMSGBOXES
+- name: Install fluentd
+  win_shell: gem install fluentd
```
Comment (ravisantoshgudimetla) [inline ansible/tasks/logging/install_fluentd.yaml]: Where do we specify the version of fluentd that we want?

File: ansible/tasks/logging/install_fluentd.yaml (hunk 5 above)
Comment (suhanime) [reply ansible/tasks/logging/install_fluentd.yaml]: I will register the variable here instead.

File: ansible/tasks/logging/main.yml (hunk 4 above)
Comment (suhanime) [reply ansible/tasks/logging/main.yml]: True. Will remove it

File: ansible/tasks/logging/install_fluentd.yaml [hunk 10]
```diff
@@ -0,0 +1,24 @@
+---
+- name: Create temporary directory on windows host
+  win_tempfile:
+    state: directory
+  register: tempdir_win
+- name: Fetch ruby devkit exe
```
Comment (ravisantoshgudimetla) [inline ansible/tasks/logging/install_fluentd.yaml]: nit: Don't address this in PR. I see that you're downloading the file onto the Windows host, while it's not a problem, you've to download it onto every host where we want the package.

File: ansible/tasks/logging/main.yml [hunk 11]
```diff
@@ -0,0 +1,16 @@
+---
+- hosts: windows
+  gather_facts: no
+  tasks:
+    - name: Ping test
+      win_ping:
+- hosts: windows
+  vars:
+    ruby_version: "RubyInstaller-2.4.7-1/rubyinstaller-devkit-2.4.7-1-x86.exe"
```
Comment (ravisantoshgudimetla) [inline ansible/tasks/logging/main.yml]: Do you want to match the versions of fluentd and ruby from linux host running in the cluster or would like to do that in a separate PR?

File: ansible/tasks/logging/main.yml [hunk 12]
```diff
@@ -0,0 +1,16 @@
+---
+- hosts: windows
```
Comment (ravisantoshgudimetla) [inline ansible/tasks/logging/main.yml]: How do you plan to test this?

File: ansible/install_td_agent.yaml (hunk 1 above)
Comment (suhanime) [reply ansible/install_td_agent.yaml]: As a part of squash, this yaml file would be dropped as it is not needed

File: ansible/tasks/logging/install_fluentd.yaml (hunk 6 above)
Comment (suhanime) [reply ansible/tasks/logging/install_fluentd.yaml]: @ravisantoshgudimetla `creates_path` does not control the installation directory. That can be specified as a separate argument in the `arguments` section of `win_package` module.

`creates_path` has the definition of `Will check the existence of the path specified and use the result to determine whether the package is already installed.` in the official documentation, however its execution is counter-intuitive. If the path mentioned in `creates_path` is valid, the task will be skipped. So we need to be very specific about the path mentioned in 'creates_path`.
Coming back to the discussion of installation directory, if we mention installation directory as, say -`D:\\ruby`, the installation will happen there. However we cannot give `D:\\ruby` as the path in `creates_path` because this  path can (and will) continue to exist if ruby is uninstalled from the system. What the `win_package` module ideally expects here is the absolute application path of the exe when it is run from the system so it knows with absolute certainty whether the said exe is already installed on the system

File: ansible/tasks/logging/README.md (hunk 7 above)
Comment (suhanime) [reply ansible/tasks/logging/README.md]: We can, however my problem mentioned in [comment above](https://github.com/openshift/windows-machine-config-operator/pull/42#discussion_r331263496) brings up the issue of whether we can make version of ruby installed dynamic. Since `creates_path` is a mandatory field, with version dependency, I am not sure if we should let the user decide the version of ruby installed

File: ansible/tasks/logging/install_fluentd.yaml (hunk 9 above)
Comment (suhanime) [reply ansible/tasks/logging/install_fluentd.yaml]: Right now we do not, so it is fetching the latest stable version of fluentd (currently 1.7.3). We can pin the version here - I will test and include the version mentioned on [origin aggregated logging](https://github.com/openshift/origin-aggregated-logging/commit/8ab73fbbfc5172d69ec81fb63dabc77196415ee6)

File: ansible/tasks/logging/README.md (hunk 2 above)
Comment (suhanime) [reply ansible/tasks/logging/README.md]: @sebsoto and I have decided to keep one README file for WSU and logging, so I will be removing the README file from this PR and create a separate PR for the same

File: ansible/tasks/logging/README.md (hunk 7 above)
Comment (suhanime) [reply ansible/tasks/logging/README.md]: Yes, that can be done. Supplying from command line can also override the existing variable in the file

File: ansible/tasks/logging/install_fluentd.yaml (hunk 3 above)
Comment (suhanime) [reply ansible/tasks/logging/install_fluentd.yaml]: I'm not sure we would need it here since the task will not be executed if the tempfile creation failed.

File: ansible/tasks/logging/main.yml (hunk 11 above)
Comment (suhanime) [reply ansible/tasks/logging/main.yml]: I can address this in a separate PR and make the fluentd version here accepted from command line for now

File: ansible/tasks/logging/install_fluentd.yaml (hunk 6 above)
Comment (suhanime) [reply ansible/tasks/logging/install_fluentd.yaml]: I have chosen to scrape the `win_package` module since it depends on pining the `creates_path` value, and instead tested the code with `win_shell` module, with enough conditionals to ensure safe execution.
I believe it is a better option because our code can then be agnostic of Ruby version as well as the installation directory

File: ansible/tasks/logging/install_fluentd.yaml (hunk 9 above)
Comment (suhanime) [reply ansible/tasks/logging/install_fluentd.yaml]: Right now this installs the latest version. The new commit should give the option of providing the version of fluentd installation

File: ansible/tasks/logging/main.yml (hunk 11 above)
Comment (aravindhp) [reply ansible/tasks/logging/main.yml]: What is the reason for going with the 32-bit version when the 64-bit version is available?

Comment (suhanime) [reply ansible/tasks/logging/main.yml]: @aravindhp We are using 64 bit version. This path corresponds to 32 bit ruby installer (but we are installing dev kit instead), and I forgot to change this path.

File: tools/ansible/tasks/logging/install_fluentd.yml [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+  register: install_dir_exists
+- name: Install ruby to particular directory
+  win_shell: "{{ ruby_devkit.dest }} /VERYSILENT /SUPPRESSMSGBOXES /DIR={{ install_dir }}"
+  when: install_dir is defined and install_dir_exists.stat.exists
+- name: Install ruby
+  win_shell: "{{ ruby_devkit.dest }} /VERYSILENT /SUPPRESSMSGBOXES"
+  when: install_dir is not defined or
+        (install_dir_exists is defined and install_dir_exists.stat.exists==false)
+- name: Check ruby installed
+  win_shell: "ruby --version"
```
Comment (aravindhp) [inline tools/ansible/tasks/logging/install_fluentd.yml]: This assumes that `ruby` is in the executable path and I am guessing the ruby devkit does that for you. So please  call that assumption out in a comment here. Same goes for `gem`.

Comment (aravindhp) [reply tools/ansible/tasks/logging/install_fluentd.yml]: It is not obvious to the reader without having prior knowledge

Comment (suhanime) [reply tools/ansible/tasks/logging/install_fluentd.yml]: I see what you mean. Okay.
In that case we should also call out why ruby-devkit instead of simple ruby install. Ruby devkit is a bundle of Ruby as well as MSYS2 packages, and both are required to install fluentd gem

File: tools/ansible/tasks/logging/main.yml [hunk 14]
```diff
@@ -0,0 +1,12 @@
+---
+- hosts: win
+  gather_facts: no
+  tasks:
+    - name: Ping test
```
Comment (sebsoto) [inline tools/ansible/tasks/logging/main.yml]: I'm not sure why a ping test is being done. This would be testing if ansible can reach the host, but thats already being tested by running the playbook itself.

File: tools/ansible/tasks/logging/ReadMe.md [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+- Set up an OpenShift cluster.
+- Set up a Windows instance as part of the cluster. You can use `wni` tool to set it up by following this [link](https://github.com/openshift/windows-machine-config-operator/tree/master/tools/windows-node-installer).
+- Ensure Ansible environment is set up to run the playbook.
+- Ensure a valid inventory file is set up.
+
+Run `main.yml` with the following command:
+```
+$ ansible-playbook -v -i <inventory_file> main.yml --extra-vars "ruby_fetch_url=<ruby_fetch_url>"
+```
+As a part of Fluentd installation, Ruby DevKit is required.
```
Comment (ravisantoshgudimetla) [inline tools/ansible/tasks/logging/ReadMe.md]: Shouldn't this be < 2.6?
 cc @sebsoto?

File: tools/ansible/tasks/logging/ReadMe.md [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+Sample command with URL:
+```
+$ ansible-playbook -v -i hosts main.yml --extra-vars "ruby_fetch_url=
+'https://github.com/oneclick/rubyinstaller2/releases/download/RubyInstaller-2.5.7-1/rubyinstaller-devkit-2.5.7-1-x64.exe'"
+```
+
+There are also some optional arguments which could be provided while running the playbook:
+- `install_dir` is the directory where you would like Ruby to be installed.
+This path should already exist on the system.
+- `fluentd_version` is the version of Fluentd to be installed. Version `1.5.1` is recommended.
```
Comment (ravisantoshgudimetla) [inline tools/ansible/tasks/logging/ReadMe.md]: is recommended by what? Are you telling you're matching the version of ruby installed on linux worker nodes?

File: tools/ansible/tasks/logging/ReadMe.md (hunk 15 above)
Comment (sebsoto) [reply tools/ansible/tasks/logging/ReadMe.md]: I'm not sure, I only tested on 2.4.6

File: tools/ansible/tasks/logging/ReadMe.md (hunk 16 above)
Comment (suhanime) [reply tools/ansible/tasks/logging/ReadMe.md]: I recommended this version after checking here: https://github.com/openshift/origin-aggregated-logging/tree/master/fluentd

File: tools/ansible/tasks/logging/ReadMe.md (hunk 15 above)
Comment (suhanime) [reply tools/ansible/tasks/logging/ReadMe.md]: According to https://rubyinstaller.org/add-ons/devkit.html it should be >=2.4.
Fluentd requires MSYS2, however I haven't tested with versions below 2.4 (they are bundled with MSYS1)

File: ansible/tasks/logging/README.md (hunk 8 above)
Comment (suhanime) [reply ansible/tasks/logging/README.md]: Logging into the cluster is not required for this playbook

File: ansible/tasks/logging/main.yml (hunk 12 above)
Comment (suhanime) [reply ansible/tasks/logging/main.yml]: Test plan for Ansible lives [here](https://docs.google.com/document/d/1tYvCHQ9toP-ukIDVqnjTvRpTVGLBGFeaXJtzLaw3jaM/edit)

File: tools/ansible/tasks/logging/ReadMe.md [hunk 17]
```diff
@@ -0,0 +1,35 @@
+# Logging on Windows instance
```
Comment (aravindhp) [inline tools/ansible/tasks/logging/ReadMe.md]: Please rename to `README.md` for consistancy

File: tools/ansible/tasks/logging/main.yml (hunk 14 above)
Comment (sebsoto) [reply tools/ansible/tasks/logging/main.yml]: This is acceptable to me, and I'd prefer it to an extra step. Ideally the user follows the documentation properly and doesn't reach a state where this happens. And if they do, things still fail in an easily understandable way assuming they are familiar with ansible.

My preference is removing the win_ping, but we can grab someone elses opinion as well if you think it makes sense.

Comment (aravindhp) [reply tools/ansible/tasks/logging/main.yml]: My preference would be to remove it too.

Comment (suhanime) [reply tools/ansible/tasks/logging/main.yml]: Sounds good. I'll remove it
